.gitlab-ci.yml

```diff
@@ -192,6 +215,52 @@ default:
   - cp -r .maintain/docker/substrate.Dockerfile ./artifacts/substrate/
   - sccache -s
 
+.cargo-doc-script:                 &cargo-doc-script
+  # FIXME: it fails with `RUSTDOCFLAGS="-Dwarnings"`
+  # Try to run with nightly rustc first. If failed, when use the stable rustc.
+  - (RUSTDOCFLAGS="--html-in-header $(pwd)/.maintain/rustdoc-header.html" &&
+    time cargo +nightly doc --no-deps --workspace --all-features --verbose) ||
+    { echo "Generate $BUILD_REF rustdocs failed" && exit 0; }
+
+.upsert-index-script:              &upsert-index-script
+  # generate index.html based on BUILD_REF
+  - yarn global add @jimmychu0807/index-tpl-crud
+  - '[[ "${RESET_INDEX}" == "true" ]] && index-tpl-crud init
+    "${GH_REPO:-$CI_PROJECT_NAME}" "Substrate" /tmp/index.html'
+  - '[[ "${LATEST}" == "${BUILD_REF}" ]]
+    && index-tpl-crud upsert -l /tmp/index.html ${BUILD_REF}
+    || index-tpl-crud upsert /tmp/index.html ${BUILD_REF}'
+
+.rm-index-script:                  &rm-index-script
+  - yarn global add @jimmychu0807/index-tpl-crud
+  - index-tpl-crud rm /tmp/index.html ${RM_RUSTDOC_REF}
+
+.reset-gh-info-script:             &reset-gh-info-script
+  # Reset git config
+  - rm -rf .git/config
+  - git config user.email "devops-team@parity.io"
+  - git config user.name "${GITHUB_USER}"
+  - git config remote.origin.url "${REPO_URL}"
+  - git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
+  - git fetch origin
+
+.git-force-push-rm-script:         &git-force-push-rm-script
+  # push it back to github gh-pages
+  - git add --all --force
+  - git commit -m "___Updated docs for ${CI_COMMIT_REF_NAME}___" || echo "___Nothing to commit___"
+  - git push origin gh-pages --force
+  - rm -rf .git/ ./*
+
+.test-build-rustdoc-script:        &test-build-rustdoc-script
+  # The actual check on whether we need to build rustdocs. This task will run if it is one of
+  #   branch/tag in CONT_RUSTDOC_BUILT_REFS or triggered via web with BUILD_RUSTDOC_REF set to
```

Comment:
Also relates to https://github.com/paritytech/substrate/pull/9856/files#diff-037ea159eb0a7cb0ac23b851e66bee30fb838ee8d0d99fa331a1ba65283d37f7R244

We are mirroring our repos the [following way](https://github.com/paritytech/devops/wiki/Gitlab:-Mirroring--setup,-CICD-Pipeline-for-Github-repos#important):
```
        fetch = +refs/heads/master:refs/remotes/upstream/master
        fetch = +refs/pull/*/head:refs/remotes/upstream/*
        fetch = +refs/tags/*:refs/tags/*
```

It means that only `master` branch keeps its' name, all the PRs are mirrored by their #. I.e. this one is #9856 so you can find your pipelines in https://gitlab.parity.io/parity/substrate/-/pipelines/164942 and code in https://gitlab.parity.io/parity/substrate/-/tree/9856.
This means that in `CONT_RUSTDOC_BUILT_REFS` you can mention either `master`, or tags, or PR numbers.

It works with the default `master` value, but how will this work with the other value?

Reply:
When a new pull request arrive, I don't foresee it to trigger this. It is only when a PR is merging to a certain branch (`master`), or a certain branch or tag name is pushed that we will trigger the rustdoc build.